[WIP] Escape of chars special to PowerShell in user data

DanHam opened the pull request:
This is a rough draft intended for discussion only. It follows on from the code and discussion in #5365 - namely auto escaping chars special to PowerShell in env vars and elevated usernames and passwords. I've also added the fix for #5368 as this was required for the demo template below to work OK.

Example:

```
  "provisioners": [
    {
      "type": "powershell",
      "inline": [
        "write-output \"Users still need to deal with special chars in commands themselves\"",
        "net user adminuser Super`$3cr3t /ADD",
        "net localgroup Administrators adminuser /add",
        "net user adminuser /ACTIVE:YES"
      ]
    },
    {
      "type": "powershell",
      "elevated_password": "Super$3cr3t",
      "elevated_user": "adminuser",
      "environment_vars": [
        "DOLLAR=A$Dollar",
        "BACKTICK=A`Backtick",
        "DBLQUOTE=A\"DoubleQuote",
        "SGLQUOTE=A'SingleQuote",
        "PSCODE=$(Write-Host \"PS Code in Env Vars should not be executed\")"
      ],
      "inline": [
        "write-output \"We didn't need to escape the dollar sign in the elevated password\"",
        "write-output \"DOLLAR is equal to $env:DOLLAR\"",
        "write-output \"BACKTICK is equal to $env:BACKTICK\"",
        "write-output \"DBLQUOTE is equal to $env:DBLQUOTE\"",
        "write-output \"SGLQUOTE is equal to $env:SGLQUOTE\"",
        "write-output \"The PowerShell code embedded in the env var was not executed\""
      ]
    },
```

This results in the following expected (and hopefully correct!) output:

```
==> virtualbox-iso: Connected to WinRM!
==> virtualbox-iso: Uploading VirtualBox version info (5.1.24)
==> virtualbox-iso: Provisioning with Powershell...
==> virtualbox-iso: Provisioning with powershell script: /var/folders/15/d0f7gdg13rnd1cxp7tgmr55c0000gn/T/packer-powershell-provisioner329677294
    virtualbox-iso: Users still need to deal with special chars in commands themselves
    virtualbox-iso: The command completed successfully.
    virtualbox-iso:
    virtualbox-iso: The command completed successfully.
    virtualbox-iso:
    virtualbox-iso: The command completed successfully.
    virtualbox-iso:
==> virtualbox-iso: Provisioning with Powershell...
==> virtualbox-iso: Provisioning with powershell script: /var/folders/15/d0f7gdg13rnd1cxp7tgmr55c0000gn/T/packer-powershell-provisioner232935784
    virtualbox-iso: We didn't need to escape the dollar sign in the elevated password
    virtualbox-iso: DOLLAR is equal to A$Dollar
    virtualbox-iso: BACKTICK is equal to A`Backtick
    virtualbox-iso: DBLQUOTE is equal to A"DoubleQuote
    virtualbox-iso: SGLQUOTE is equal to A'SingleQuote
    virtualbox-iso: The PowerShell code embedded in the env var was not executed
```

Notes:

* I've taken out the commit in #5365 that changed the env var formats to use single quotes
* I've based (blatantly ripped off!!!) the code that escapes chars special to PowerShell on the Go code that escapes chars special to XML - see the ```func escapeText``` [here](https://golang.org/src/encoding/xml/xml.go)
* I'm not sure about the character range used in ```func isInCharacterRange```. Clearly things work OK as is, but I wondered if that needed amending in some way to be 100% compatible/applicable to PowerShell.
* This should 'fix' #5258.
* I've expanded the tests to check env vars with chars special to PS are dealt with properly. Do we need to also test that elevated username and elevated passwords are escaped properly as well? I would probably need some help with this
* Users that have escaped chars special to PowerShell in their env vars or elevated_user or elevated_password will now need to remove the escape e.g.

```
      "elevated_user": "Administrator",
      "elevated_password": "Super`$3cr3t!"
```

Will just need to become:

```
      "elevated_user": "Administrator",
      "elevated_password": "Super$3cr3t!"
```

Closes #5471

SwampDragons commented:
I really like this start and think this is the right way to go.

SwampDragons commented:
We'll have to write a fixer for this before we merge it.

DanHam commented:
Just for now, I've rebased and fixed up to align this PR with changes to #5365 and merge of #5369. This will help (me) keep things straight...

@SwampDragons:

> I really like this start and think this is the right way to go.

Neat! Thanks!

> We'll have to write a fixer for this before we merge it.

OK - so good start but wrong place? 

I've been taking a look at the other fixers that are under fix/. I wasn't aware of these before but, if I'm understanding things correctly, I seem to have made a good start... but made it in the wrong place!

Creating a new fixer to do this (rather than trying to do this in the provisioner itself) looks like fun. I'll see if I can get something working and resubmit for review... If not I may be asking for some help!

SwampDragons commented:
No, I think you're in the right place. Escaping here is good.  The fixers are basically ways of making it easy for people to update their configs to protect against backwards-incompatible changes.  So if someone's variable is "Dollar$ign" then they'll have it in their config as "Dollar`$ign" but after our change they won't need the backtick.  So we need a fixer that un-escapes the variables in their config, that they can run one time to fix the config.

DanHam commented:
> ...we need a fixer that un-escapes the variables in their config

@SwampDragons I think I posted the edit to my comment around the same time as you posted the full explanation... but never mind! I now understand what you want from the fixer.

SwampDragons commented:
Hah good!  So many edits, so little time.  You're exactly right -- the fixers are the functionality behind `packer fix`.  Let me know if you want a hand, but several of the fixers in the repo are simple and very copy-pasteable

DanHam commented:
@SwampDragons Would you mind letting me know what you think of this new approach? 

It always seems to be the way, but I think I overcomplicated this initially and have now found what I think is a much better/simpler approach. 

[This approach](https://github.com/hashicorp/packer/pull/5376/commits/c9b6c12daa9683c61e71c827d1401f1753e969f8), to my mind is much simpler than my [earlier attempt](https://github.com/hashicorp/packer/pull/5376/commits/53aefe744bcd74ee6ac866f764eafe9e7f507d80). 

It will still work the same way from the users perspective and should produce the same output given the sample provisioner section in the template above.

I've rebased so this is up to date. However, it will obviously need to be rebased again to lose/use whatever approach you think is best.

Also, I still haven't been able to find time to look at the fixer we will need yet :/ ...

DanHam commented:
Just a quick note so I remember that we will need to update the example docs [here](https://github.com/hashicorp/packer/blob/master/website/source/intro/getting-started/build-image.html.md#a-windows-example)...

SwampDragons commented:
I changed the milestone to our next minor release since this will break templates without the fixer.

I'll read through your new approach today and let you know what I think :)

DanHam commented:
@SwampDragons 

>I'll read through your new approach today and let you know what I think

I realise you've been really busy! However, did you manage to find time to take a look at this?

I've (finally) managed to have a go at the fixer for this. I've been using [this template](https://github.com/DanHam/packer-testing/tree/master/aws/test-ps-escapes) for testing.

Note that the fixer cannot auto fix escapes in env vars or elevated username/passwords that are expanded from user variables - see VAR4 in the final provisioner of the example template. I think that this would prove to be quite a challenge!

I think we need #5515 to be merged prior to this PR for escaping of double quotes to work. At present I think something wierd goes on with the escape for json and then the escape for powershell e.g. If we had something like:

```
    {
      "type": "powershell",
      "environment_vars": [
        "VAR1=A`$Dollar",
        "VAR2=A``Backtick",
        "VAR3=A`'SingleQuote",
        "VAR4=A`\"DoubleQuote"
        ...
    }
```

...VAR4 would cause us problems at present! While I haven't tested this, my suspicion is that VAR4 would cause an error with the current version of Packer as well. I'm *fairly* confident #5515 will fix this; The problem seems to be caused because the env vars are embedded in the command string. Again, I haven't had time to test this. 

I see you have tagged this PR and #5515 for the v1.2.0 release. I agree that this seems like the best option; However, it does pose a problem as merging #5515 would then require we fix up this PR - the changes #5515 would cause a conflict with this PR. How do you suggest we solve? Should we put all the code into 1 PR?

Outstanding (note to myself): 

- [ ] Change example Windows build docs - template and sample script [here](https://github.com/DanHam/packer/blob/master/website/source/intro/getting-started/build-image.html.md)
- [ ] Tests for new fixer
- [ ] Update docs

DanHam commented:
@SwampDragons 

> ...VAR4 would cause us problems at present! While I haven't tested this, my suspicion is that VAR4 would cause an error with the current version of Packer as well. I'm fairly confident #5515 will fix this; The problem seems to be caused because the env vars are embedded in the command string. Again, I haven't had time to test this.

So, just tested all this with #5515 and this PR merged into one branch. Good news is that it fixes the issue with double quotes as we suspected it would:

```
    amazon-ebs: Fixer removed escape for dollar in VAR1:      A$Dollar
    amazon-ebs: Fixer removed escape for backtick in VAR2:    A`Backtick
    amazon-ebs: Fixer removed escape for singlequote in VAR3: A'SingleQuote
    amazon-ebs: Fixer removed escape for doublequote in VAR4: A"DoubleQuote
    amazon-ebs: Fixer does NOT fix dollar in VAR5 as this is expanded from user variable - we end up with extra escapes for backtick and dollar
```

See the ps-all branch on my Packer fork [HERE](https://github.com/DanHam/packer/tree/ps-all) and the Packer templates I used to test with [HERE](https://github.com/DanHam/packer-testing/tree/master/aws/test-ps-all).

My feeling is that we should close #5515 and this PR and proceed with the ps-all branch (e.g. with #5515 and this PR combined)...
What do you think?

SwampDragons commented:
Oh, yeah, I read it that day and didn't ever respond.  I think it looks good.

Good news that the combined branch fixes VAR4.  I'm happy to have you roll them together and merge them into master at the same time.

I think we're just going to have to make sure we communicate really well about this since you're right that the fixer won't fix the environment variables.  When we write to the mailing list we'll make sure that we call out this backwards incompatibility.  I'll also make sure the note in the changelog is detailed enough that people can fix their vars.

SwampDragons commented:
@DanHam Did you ever make a ps-all branch with all your changes together?  We're coming up on our 1.2.0 release which means this backwards-incompatable change is allowed in.

SwampDragons commented:
Merged 5515; just rebased and force-pushed this one on top of master, including 5515.  I'm going to do a few tests and take a few notes, then merge.

DanHam commented:
@SwampDragons Hello! I did create a ps-all branch - I've got that ready(ish) on my clone of packer. Really sorry for the long period of no action on this... However, I've moved house, had no connection to the internet for about 6 weeks and then had a bout of opthalmic shingles which knocked me out for about a month! Thankfully I'm now back on my feet and life is getting back to normal again...

I'll take a look at this ASAP and check we're all good for to getting this into 1.2.0.

Would you be ok with reverting this merge and 5515? Things might be cleaner if we can work straight off the ps-all branch. I had some initial work I reverted on this PR as I found a simpler way to do things... I cleaned all of that up in the ps-all branch...

SwampDragons commented:
Hey Dan!  No worries about the slow response; I should have given you a little more time before charging ahead.  I've just reverted both this merge and 5515.

DanHam commented:
Great! I'll get to work on this over the next few days! Have a great weekend!

SwampDragons commented:
Sounds good -- we want to do the 1.2.0 release next week, so please let me know if you think you're gonna miss the deadline.
